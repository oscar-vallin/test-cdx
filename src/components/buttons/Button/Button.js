import PropTypes from 'prop-types';
import { StyledButton } from './Button.styles';

const Button = ({
  id = '__Button',
  text,
  children,
  variant = 'secondary',
  disabled = false,
  onClick,
  block = false,
  ...props
}) => {
  return (
    <StyledButton id={id} variant={variant} disabled={disabled} onClick={onClick} block={block} {...props}>
      {text || children}
    </StyledButton>
  );
};

Button.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
  variant: PropTypes.string,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  block: PropTypes.bool,
};

export { Button };
