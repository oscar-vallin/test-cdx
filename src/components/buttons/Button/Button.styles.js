import styled from 'styled-components';
import { PrimaryButton } from '@fluentui/react';

const getVariant = (theme, variant) => {
  const VARIANTS = {
    primary: { background: theme.colors.themePrimary, color: '#fff' },
    secondary: { background: theme.colors.white, color: theme.colors.themePrimary },
    light: { background: 'transparent', color: theme.colors.themePrimary },
    danger: { background: 'transparent', color: theme.colors.custom.error },
    error: { background: theme.colors.custom.severeError, color: theme.colors.white },
  };

  return VARIANTS[variant];
};

export const StyledButton = styled(PrimaryButton)`
  width: ${({ block }) => (block ? '100%' : 'auto')};
  
  &,
  & + button {
    background: ${({ theme, variant }) => getVariant(theme, variant).background};
    border ${({ theme, variant }) => `1px solid ${variant === 'secondary' ? theme.colors.black : 'transparent'}`};
    color: ${({ theme, variant }) => getVariant(theme, variant).color};
    
    &:hover {
      background: ${({ theme, variant }) => getVariant(theme, variant).background};
      color: ${({ theme, variant }) => getVariant(theme, variant).color};
    }

    i {
      color: ${({ theme, variant }) => getVariant(theme, variant).color};
    }
  }

  & + button {
    & + span {
      background: ${({ theme, variant }) => getVariant(theme, variant).color};
    }
  }
`;
