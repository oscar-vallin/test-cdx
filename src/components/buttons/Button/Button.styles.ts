import styled from 'styled-components';
import { PrimaryButton } from '@fluentui/react';

const getBorderVariant = (theme, variant) => {
  return `1px solid ${variant === 'secondary' ? theme.colors.black : 'transparent'}`;
};

const getVariant = (theme, variant) => {
  const VARIANTS = {
    primary: { background: theme.colors.themePrimary, color: '#fff' },
    secondary: { background: theme.colors.white, color: theme.colors.themePrimary },
    navbar: { background: theme.colors.navbar, color: theme.colors.white },
    light: { background: 'transparent', color: theme.colors.themePrimary },
    danger: { background: 'transparent', color: theme.colors.custom.error },
    error: { background: theme.colors.custom.severeError, color: theme.colors.white },
  };

  return VARIANTS[variant] ?? VARIANTS.primary;
};

export const StyledButton = styled(PrimaryButton)<StyledButtonProps>`
  font-size: 0.875rem;
  width: ${({ block }) => (block ? '100%' : 'auto')};

  &,
  & + button {
    background: ${({ theme, variant }) => getVariant(theme, variant).background};
    border: ${({ theme, variant }) => getBorderVariant(theme, variant)};
    color: ${({ theme, variant }) => getVariant(theme, variant).color};
    white-space: nowrap;

    &:hover {
      background: ${({ theme, variant }) => getVariant(theme, variant).background};
      color: ${({ theme, variant }) => getVariant(theme, variant).color};
    }

    i {
      color: ${({ theme, variant }) => getVariant(theme, variant).color};
    }
  }

  & + button {
    & + span {
      background: ${({ theme, variant }) => getVariant(theme, variant).color};
      color: ${({ theme, variant }) => getVariant(theme, variant).color};
    }
  }

  &:disabled {
    background: ${({ theme }) => theme.colors.neutralTertiaryAlt};
  }
`;

type StyledButtonProps = {
  variant?: string;
  block?: boolean;
};
