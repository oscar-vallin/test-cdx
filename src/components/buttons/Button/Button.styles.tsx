import styled from 'styled-components';

export const StyledComponent = styled.button`
  width: 320px;
  height: 32px;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  padding: 6px 0px;

  /* Theme / Primary */
  background: #0078d4;
  border-radius: 2px;
  color: #fefefe;
`;
