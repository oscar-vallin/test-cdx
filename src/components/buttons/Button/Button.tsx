import React from 'react';
import { StyledComponent } from './Button.styles';

const DEFAULT_FUNCTION = () => console.log('Click');

export const Button = ({ children = <></>, id = '', onClick = DEFAULT_FUNCTION }) => {
  return (
    <StyledComponent id={`Button__${id}`} onClick={onClick}>
      {children}
    </StyledComponent>
  );
};
