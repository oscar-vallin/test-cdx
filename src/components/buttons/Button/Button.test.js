import React from 'react';
import { shallow } from 'enzyme';
import { Button } from './Button';

describe('Button', () => {
  const mockFn = jest.fn();

  it('Should be defined', () => {
    expect(Button).toBeDefined();
  });

  it('Should render correctly', () => {
    const tree = shallow(<Button onClick={mockFn}>Testing Button</Button>);
    expect(tree).toMatchSnapshot();
  });

  it('Should have a click function', () => {
    const tree = shallow(<Button onClick={mockFn}>Testing Button</Button>);
    tree.simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });
});
