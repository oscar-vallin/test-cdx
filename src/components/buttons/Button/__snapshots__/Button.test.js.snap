// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Testing Unit... Should render correctly 1`] = `
<Styled(CustomizedPrimaryButton)
  disabled={false}
  id="__Button"
  onClick={[MockFunction]}
  variant="secondary"
>
  Testing Button
</Styled(CustomizedPrimaryButton)>
`;
